src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+{
+    /// <summary>
+    /// Provider that computes display order of tree items based on input ordering of
+    /// evaluated includes from the project file.
+    /// </summary>
+    internal class TreeItemOrderPropertyProvider : IProjectTreePropertiesProvider
+    {
+        private readonly char[] _separators = new char[]
+        {
+            Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar
+        };
+
+        public IReadOnlyCollection<ProjectItemIdentity> OrderedItems { get; }
+
+        public ImmutableHashSet<string> AllowedItemTypes { get; }
+
+        public Dictionary<string, int> DisplayOrderMap { get; } = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+
+        public TreeItemOrderPropertyProvider(IReadOnlyCollection<ProjectItemIdentity> orderedItems)
+        {
+            OrderedItems = orderedItems;
+
+            AllowedItemTypes = OrderedItems.Select(p => p.ItemType).ToImmutableHashSet();
+
+            ComputeIndices(orderedItems);
+        }
+
+        /// <summary>
+        /// Preorder folders and items that are provided as ordered evaluated includes
+        /// </summary>
+        /// <param name="orderedItems">ordered evaluated includes</param>
+        private void ComputeIndices(IReadOnlyCollection<ProjectItemIdentity> orderedItems)
```

Comment:
>ComputeIndices [](start = 21, length = 14)

I don't get what this method is doing actually .. I assume it makes sense to talk about the order for items or folders located at the same level, for example if project contains:

u\a\w\d.cs
a\b\x.cs
a\b\a.cs
_c\d\f\1.cs

here you only care about order of a\b\x.cs
and a\b\a.cs, since they are in the same folder and user put x above a. Everything else should not be ordered isn't it? 

In your implementation you would put folder u above folder a and _c , but user did not say anything about lower level folders - isn't it? If user wants to order folders explicitly, he would add explicit project items:

<Folder Include="u\" />
<Folder Include="a\" />

Or i am missing something for F#?

Reply:
Yes, in F# users do not explicitly order the folders. So if they have a bunch of entries in u\... followed by a bunch of entries in a\... they would expect 'u' to come before 'a'. If they mix the entries, that behavior is undefined (and disallowed), so my implementation just picks the order of the first one to occur. Note that during CalculatePropertyValues below, we do not have the full evaluated include anymore, just the item name/type. So if the include is "Orders/Order.fs", we get two separate items for "Order" and "Order.fs". We can improve this implementation by modifying the data flow to return the original include with each item, so we can distinguish between duplicates.

---
In reply to: [148054784](https://github.com/dotnet/project-system/pull/2930#discussion_r148054784) [](ancestors = 148054784)

Reply:
Also /cc @brettfo @KevinRansom to comment on this. My understanding is the classic project system is limited in its handling of folders, and most F# projects leave all files at the root level. We could avoid some of this complexity by also ignoring folders, but this implementation does provide a default that will work with the paradigms I understand to be in use.

---
In reply to: [148099504](https://github.com/dotnet/project-system/pull/2930#discussion_r148099504) [](ancestors = 148099504,148054784)

Reply:
I thought we agreed to ignore folder ordering, with regard to eachother, and instead just keep them at the top, with their contents ordered correctly. This was required due to limitations in CPS that will be investigated at a future date.

Reply:
@tannergooding: Folder ordering (as specified in the `.fsproj`) is required because ultimately the compiler doesn't care about folder structure, it only cares that certain files are given to it ahead of others, meaning you can't just consider `a.fs` to be the file name, but rather `x\y\z\a.fs`; the folder is simply a display mechanism.

@natidea: Many projects leave `.fs` files at the root, but not all, including `FSharp.Core` (currently `FSharp.Core` is a desktop project, but we're converting it to an SDK project as soon as possible.)

Consider this example; neither the folders nor the contents are sorted and the files as listed in `FSharp.Core.fsproj` are:

``` xml
...
<Compile Include="Primitives\prim-types-prelude.fsi" />
<Compile Include="Primitives\prim-types-prelude.fs" />
<Compile Include="Primitives\prim-types.fsi" />
<Compile Include="Primitives\prim-types.fs" />
<Compile Include="Collections\local.fsi" />
<Compile Include="Collections\local.fs" />
<Compile Include="Collections\array2.fsi" />
<Compile Include="Collections\array2.fs" />
...
```

And it currently displays in VS like so:
![image](https://user-images.githubusercontent.com/926281/32301712-4f6f7b7e-bf1c-11e7-87d1-bf13f64fe839.png)